### PR TITLE
[EntraWorkloadIAM] Use config protected settings

### PR DESCRIPTION
**Pull request by AntonioND**
Instead of using regular configuration settings for the join token, use protected configuration settings.

This is an example of the output when the extension is installed with the changes in the PR (the value of the join token isn't shown):

image

image

---

### General Guidelines

- [X] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [X] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [X] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

### CODE REVIEW

File: src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py [hunk 1]
```diff
@@ -65,24 +84,25 @@ def Create(self, cmd, client, resource_group_name, cluster_name, name, cluster_t
...
+        config_settings = settings_dict_to_lowercase(configuration_settings)
+        config_protected_settings = settings_dict_to_lowercase(configuration_protected_settings)
 
         # Get user configuration values and remove them from the dictionary so that they aren't
         # passed to the Helm chart
         trust_domain = config_settings.pop(CONFIG_SETTINGS_USER_TRUST_DOMAIN, None)
         tenant_id = config_settings.pop(CONFIG_SETTINGS_USER_TENANT_ID, None)
         local_authority = config_settings.pop(CONFIG_SETTINGS_USER_LOCAL_AUTHORITY, None)
+
+        # The join token can also be provided by the user. However, we must prevent them from
```
Comment (dafang982) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py:97]: I wonder whether we should stop allowing it being passed in as config_settings and only allowing config_protected_settings. Thoughts?

Comment (dafang982) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py:97]: I like your approach to completely preventing users passing join token in as unsafe config settings.

**Comment by dafang982** [comment]
Great PR. Thanks Antonio. Can you attach screenshot for the manual test that shows the updated command successfully deploys the Arc extension please?

**Review by dafang982 — APPROVED** [review]
LGTM

**Comment by AntonioND** [comment]
> Great PR. Thanks Antonio. Can you attach screenshot for the manual test that shows the updated command successfully deploys the Arc extension please?

Done!

**Comment by AntonioND** [comment]
Hi, @bavneetsingh16, @deeksha345, @anagg929, can we get some feedback on this PR?

File: src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py [hunk 2]
```diff
@@ -65,24 +84,25 @@ def Create(self, cmd, client, resource_group_name, cluster_name, name, cluster_t
...
+        # passing it as a regular configuration setting, it must be passed as a protected setting.
         join_token = config_settings.pop(CONFIG_SETTINGS_USER_JOIN_TOKEN, None)
+        if join_token is not None:
+            raise InvalidArgumentValueError(
+                "The join token must be provided with --config-protected-settings, not "
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py]: >join token [](http://example.com/codeflow?start=21&length=10)

Consider using 'jointoken' - same as how the user would give this parameter - so it is easily recognizable.

File: src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py [hunk 3]
```diff
@@ -106,9 +126,14 @@ def Create(self, cmd, client, resource_group_name, cluster_name, name, cluster_t
...
-        configuration_settings[CONFIG_SETTINGS_HELM_JOIN_TOKEN] = join_token
 
-        logger.debug("Configuration settings value for Helm: %s" % str(configuration_settings))
+        if configuration_protected_settings is None:
+            configuration_protected_settings = dict()
+        configuration_protected_settings[CONFIG_SETTINGS_HELM_JOIN_TOKEN] = join_token
+
+        logger.debug("Configuration settings: %s" % str(configuration_settings))
+        logger.debug("Configuration protected settings value for Helm: %s" %
+                     str(configuration_protected_settings))
```
Comment (NarayanThiru) [inline src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py]: These are protectedSettings, right?  Is it necessary to write this?  Can you just write the keys, without the values - just to be safe.

**Review by NarayanThiru — CHANGES_REQUESTED** [review]
@AntonioND , I left a couple of comments.  Can you pl. review them?

File: src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py (hunk 1 above)
Comment (NarayanThiru) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py:97]: Agree.  Is there any other known setting that you may want to validate as well?

Comment (AntonioND) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py:97]: I think this is the only setting that we don't want to expose so clearly. Checking the others doesn't make much sense, I believe.

File: src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py (hunk 3 above)
Comment (AntonioND) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py]: Oops, right! Thanks for noticing. It wouldn't be a huge deal if it was printed, but yeah, if we go through this effort to have the setting as a protected setting we might as well hide it from here.

File: src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py (hunk 2 above)
Comment (AntonioND) [reply src/k8s-extension/azext_k8s_extension/partner_extensions/EntraWorkloadIAM.py]: Done! I've also changed it in another user error message.

**Comment by AntonioND** [comment]
> @AntonioND , I left a couple of comments. Can you pl. review them?

Thanks! I've taken care of them now. Sorry for the delay, I was away on holiday.

**Comment by NarayanThiru** [comment]
No worries, @AntonioND.  Looks good, and I am signing off.

**Review by NarayanThiru — APPROVED** [review]

**Comment by AntonioND** [comment]
Any news about this, @NarayanThiru?

**Comment by NarayanThiru** [comment]
@bavneetsingh16 , can you pl. help merge this for @AntonioND ?

**Pull request merged**